### ISSUE TITLE
Optimize operation propagation

### ISSUE THREAD

**Issue by damip** [massalabs/massa#2283]
# Rationale

Naked (out-of-block) operations represent the bulk of the network activity under heavy load, as in the ternet currently.
We also observe that the network has no latency problems.

We need to reduce bandwidth usage by propagating operations in a way similar to what we do for blocks: send a small advert to everyone around us except nodes that already have it, and send the full operation only to who ask for it.

This will massively reduce bandwidth usage (~x10) which is desirable, and only slightly increase the latency of operation propagation which is non-critical because operation inclusion time is t0/2 = 8 seconds on average.

#  How we do it

## Setup

Have a `asking_ops: HashMap<OperationId, (NodeId, Instant)>` field in Protocol indicating to which nodes we are asking which operations and when we started doing so.

## Asking process (ask_ids)

This will figure out the best way to ask for a list of op IDs (ask_ids).

* remove all IDs that are in asking_ops from ask_ids
* remove all operation IDs we know about from ask_ids
* remove all IDs that no node knows about from ask_ids
* while ask_ids is not empty:
  * compute the intersection inter_ids of each node's operation knowledge with ask_ids
  * choose the node with the largest inter_ids to batch as much as possible (TODO also sort by how much stuff that node is already being asked for to prevent flooding ? By how recently the intersection knowledge was announced because because ops could have been honestly dropped by nodes since they announced them ?):
    * send a AskOperations(inter_ids) message to the node 
    * insert inter_ids entries in ask_ids for that node and time now()
    * remove inter_ids from ask_ids

## asking_ops management

* if a node disconnects, if it had an entries in asking_ops:
  * remove all entries matching this node ID and gather all the operation ids into a ask_ids hashset, then ask for them using the Asking process above
* if an element in asking_ops times out according to a config-defined timeout
  * ban the node and run the "node disconnects" process above

## Operations message

Remove the Operations messasge

## Advertise operations

Create the message::AdvertiseOperations message to advertise a list of operation IDs to a node:
```rust
AdvertiseOperations(Vec<OperationId>)
```
This message can contain up to `massa_network::settings::max_operations_per_message` IDs. If a request needs more elements, it must be split into multiple messages.

This messsage is sent to nodes not already knowing about an operation that we just learned about and received in full.

When we receive a AdvertiseOperations message, we mark the sender node as knowing those operations (like we do for blocks when we receive their header), then we run the Asking process for the received operations list.

## Ask for operations

Create the message::AskOperations message to ask a node for a list of operations by ID. 
```rust
AskOperations(Vec<OperationId>)
```
This message can contain up to `massa_network::settings::max_operations_per_message` IDs.

Whenever a node receives an AskOperations message:
* mark the sender node as not knowing those operations
* send a `OperationSearchResult(HashMap<OperationId, Option<Operation>>)` message listing the result of looking for those operations

## Operation search result

```rust
OperationSearchResult(HashMap<OperationId, Option<Operation>>)  // None if op not found
```

when receiving `OperationSearchResult(HashMap<OperationId, Option<Operation>>)`:
* for all found operations:
    * mark the sender node as knowing them now()
    * remove them from asking_ops
    * process the ops
* for all not-found operations:
    * mark the sender as not knowing them
    * remove them from asking_ops
    * then we run the Asking process for those operations

**Comment by damip** [comment]
TO BE DISCUSSED

**Comment by AureliaDolo** [comment]
It may be worth refactor the way we manage blocks and header in protocol into some high level mechanism that would work for  operations too.

But I'm not sure how that would interact with the Great Refactoring

**Comment by damip** [comment]
> It may be worth refactor the way we manage blocks and header in protocol into some high level mechanism that would work for operations too.
> 
> But I'm not sure how that would interact with the Great Refactoring

Unfortunately the mechanisms are a bit different:
* for blocks, we want stuff that was announced, as well as dependencies, and we ask all nodes for them
* block propagation time is critical
* for ops, we only want announced ops, and only from the nodes that announced them
* ops propagation time is not critical
* headers are small fraction of the bandwidth
* ops are large fraction of the bandwidth
* there are 2 blocks per second and 10000+ tx/s so we ave constraints on algorithm efficiency and batching

**Comment by sebastien-forestier** [comment]
Another way of doing it, with a focus on saving cpu time.

### General Idea 
When we receive a batch of operation ids from a node, we ask it the ones that we did not receive yet and did not ask yet, and we postpone to 200ms in the future the processing of the operation ids that we did not receive yet but we did ask in the last 200m.

### Assumptions
* We have a function `is_op_received(op_id)` returning `True` if the node already received the full operation with id `op_id`.

### Constants
* `operation_id_size` = 128 bits
* `op_batch_proc_period` = 200 ms
* `max_op_batch_per_sec_per_node` = 6 batches
* `max_op_per_batch` = 500 operation ids
* `op_batch_buf_capacity` = 1000 items
* `asked_cache_lifetime` = 100 sec

### Data Structures
* `asked_ops: HashMap<OperationId, (Instant, HashSet<NodeId>)>` temporarily records to which nodes we asked each operation and the last instant we asked for the operation.
* `op_batch_buf: Buffer<(Instant, NodeId, HashSet<OperationId>)>` with capacity `op_batch_buf_capacity` queues the batches of operation ids to be processed, with information on the node id from which we received the batch and the time when it should be processed.

### Threads/Triggers
* when we receive a batch `op_id_batch` of operation ids from node `node_id`, we `process_op_batch(op_id_batch, node_id)`.
* while the first element `(op_id_batch, node_id, timestamp)` of `op_batch_buf` has a timestamp before now, we pop it and `process_op_batch(op_id_batch, node_id)`, then we wait for the next timestamp.
* every `asked_cache_lifetime`, we empty `asked_ops`.

### Algorithms
```python
def process_op_batch(op_batch, node_id):
    ask_set = void HashSet<OperationId>
    future_set = void HashSet<OperationId>
    for op_id in op_batch:
        if not is_op_received(op_id):
            if (op_id not in asked_ops) or (node_id not in asked_ops(op_id)[1]):
                if (op_id not in asked_ops) or (asked_ops(op_id)[0] < now - op_batch_proc_period:
                    ask_set.add(op_id)
                    asked_ops(op_id)[0] = now
                    asked_ops(op_id)[1].add(node_id)
                else:
                    future_set.add(op_id)
    if op_batch_buf is not full:
        op_batch_buf.push(now+op_batch_proc_period, node_id, future_set)
    ask ask_set to node_id
```
### Notes
* the node should send operation ids to other nodes by batch, for instance with a 200ms buffer.
* later, we could deprioritize operation processing/broadcasting vs block consensus, e.g. when we're late on VM processing w.r.t. the cursor, we can postpone the processing of the received batch by pushing it to the buffer if the buffer is not full and drop it otherwise, and postpone the processing of items of the buffer.

**Comment by gterzian** [comment]
@massalabs/core-team 

I think we can do this exactly like we currently already do for blocks. 

Main points:

- From "the point of view of the local Node", there is a [`block_wishlist`](https://github.com/massalabs/massa/blob/be86a1e8d3f19f2e6bbda019fb4937392a41990e/massa-protocol-worker/src/protocol_worker.rs#L306), which is processed by way of [`update_ask_block`](https://github.com/massalabs/massa/blob/be86a1e8d3f19f2e6bbda019fb4937392a41990e/massa-protocol-worker/src/protocol_worker.rs#L701)
- For "the other nodes to which the local node is connected", there is [`NodeInfo.wanted_blocks`](https://github.com/massalabs/massa/blob/be86a1e8d3f19f2e6bbda019fb4937392a41990e/massa-protocol-worker/src/protocol_worker.rs#L117). 

See also the [test suite](https://github.com/massalabs/massa/blob/be86a1e8d3f19f2e6bbda019fb4937392a41990e/massa-protocol-worker/src/tests/ask_block_scenarios.rs) that comes with it.

@sebastien-forestier I believe `update_ask_block` contains a timing mechanism similar to the one your propose(and if not it could be added to it). 

I think this could initially be a copy/paste and search/replace type of work. 

Later we can probably do something to some some sort of `Wishlist<T>` object...

**Comment by gterzian** [comment]
Some other notes on the implementations, related to communication between components(as opposed to the internal logic of `protocol_worker` and the various network messages as already specced above):

- The current use by pool of `propagate_operations` would have to be replaced by a "pull" workflow coming from `protocol`. This would use a new `ProtocolPoolEvent::GetOperations`, and coud re-use the current `ProtocolCommand::PropagateOperations`. 

In other words, `pool` would only send a `ProtocolCommand::PropagateOperations` in response to receiving a `ProtocolPoolEvent::GetOperations` message.

**Comment by gterzian** [comment]
@massalabs/core-team @sebastien-forestier Re-opening this one se we can discuss how we can improve the spec(if anything appears to have remained unclear), and perhaps also how we can evolve the spec as we change the code(for example https://github.com/massalabs/massa/issues/2630)

**Comment by damip** [comment]
It is now pretty well documented and established. Closing for now
